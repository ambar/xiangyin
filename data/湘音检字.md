# 湘音检字

## 修正

- 去除了重复的行（3K+）
- 「攡」应当在字甲

## 问题

文件默认包含了一些简体/异体对应，但不全面，一些例外：

```text
弄 衖
纤 纖
减 減
们 們
础 礎
迁 遷
团 團
丛 叢
导 導
头 頭
触 觸
沣 灃
尝 嘗
```

## 参考

- [《湘音检字》与长沙方言 - 鲍厚星](https://www.zhihu.com/xen/market/remix/paid_magazine/1395144375069016064)
- [《湘音检字》及其反映的三套语音系统 - 田范芬](https://www.zhihu.com/xen/market/remix/paid_magazine/1395738931791511552)
- [两本长沙方言韵书:《训诂谐音》和《湘音检字》 - 田范芬](https://www.zhihu.com/xen/market/remix/paid_magazine/1394207942187569152)
