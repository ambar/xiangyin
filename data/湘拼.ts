import * as m3 from '../data/汉语方音字汇.meta'
import * as m1 from '../data/湘音检字.meta'
import * as m2 from '../data/长沙话音档.meta'

// TODO: 改 union
type XPAInitial = string
type XPAFinal = string
type InitialRow = [
  m1.Initial | null,
  m2.Initial | null,
  m3.Initial | null,
  XPAInitial
]
type FinalRow = [
  m1.Final | null,
  m2.Final | null,
  m3.Final | null,
  //
  XPAFinal
]

export const InitialConfig: InitialRow[] = [
  ['', '', '', ''],
  ['p', 'p', 'p', 'b'],
  ['pʰ', 'pʰ', 'pʰ', 'p'],
  ['m', 'm', 'm', 'm'],
  ['s', 's', 's', 's'],
  ['ts', 'ts', 'ts', 'z'],
  ['tsʰ', 'tsʰ', 'tsʰ', 'c'],
  ['t', 't', 't', 'd'],
  ['tʰ', 'tʰ', 'tʰ', 't'],
  ['ȵ', 'ȵ', 'ȵ', 'gn'],
  // sh -> x
  ['ɕ', 'ɕ', 'ɕ', 'x'],
  ['k', 'k', 'k', 'g'],
  ['kʰ', 'kʰ', 'kʰ', 'k'],
  // zh -> j
  ['tɕ', 'tɕ', 'tɕ', 'j'],
  // ch -> q
  ['tɕʰ', 'tɕʰ', 'tɕʰ', 'q'],
  ['x', 'x', 'x', 'h'],
  ['ŋ', 'ŋ', 'ŋ', 'ng'],
  // 李
  ['l̃', null, null, 'l'],
  [null, 'l', 'l', 'l'],
  // 发
  ['ɸ', null, null, 'f'],
  [null, 'f', 'f', 'f'],
  //  紙長
  ['tʃ', null, null, 'zh'],
  [null, 'tʂ', null, 'zh'],
  // 車尺
  ['tʃʰ', null, null, 'ch'],
  [null, 'tʂʰ', null, 'ch'],
  // 惹
  ['ɹ', null, null, 'r'],
  [null, 'ʐ', null, 'r'],
  [null, null, 'z', 'r'],
  // 诗尸
  ['ʃ', null, null, 'sh'],
  [null, 'ʂ', null, 'sh'],
]

export const FinalsConfig: FinalRow[] = [
  // '' -> i
  ['ɿ', 'ɿ', 'ɿ', 'i'],
  [null, 'ʅ', null, 'i'],

  ['ɑ', null, null, 'a'],
  [null, 'a', 'a', 'a'],

  ['i', 'i', 'i', 'i'],
  ['u', 'u', 'u', 'u'],
  ['y', 'y', 'y', 'y'],

  ['iɑ', null, null, 'ia'],
  [null, 'ia', 'ia', 'ia'],

  ['yɑ', null, null, 'ya'],
  [null, 'ya', 'ya', 'ya'],

  ['uɑ', null, null, 'ua'],
  [null, 'ua', 'ua', 'ua'],

  ['ɘ', null, null, 'e'],
  ['ie', 'ie', 'ie', 'ie'],
  ['e', null, null, 'e'],
  [null, 'ə', null, 'e'],
  [null, null, 'ɤ', 'e'],

  ['o', 'o', 'o', 'o'],
  ['io', 'io', 'io', 'io'],
  ['ai', 'ai', 'ai', 'ai'],
  ['uai', 'uai', 'uai', 'uai'],
  ['yai', 'yai', 'yai', 'yai'],

  ['uɘ', null, null, 'ue'],
  [null, 'uə', null, 'ue'],
  [null, null, 'uɤ', 'ue'],

  ['ye', 'ye', 'ye', 'ye'],
  ['yɘi', null, null, 'yei'],
  [null, 'yei', 'yei', 'yei'],

  ['uɘi', null, null, 'uei'],
  [null, 'uei', 'uei', 'uei'],

  ['ɘi', null, null, 'ei'],
  [null, 'ei', 'ei', 'ei'],

  ['au', 'au', 'au', 'au'],
  // ['x', null, null],
  ['iau', 'iau', 'iau', 'iau'],
  ['əu', 'əu', 'əu', 'ou'],
  ['iəu', 'iəu', 'iəu', 'iou'],
  ['ã', null, null, 'an'],
  [null, 'an', 'an', 'an'],

  ['iã', null, null, 'ian'],
  [null, 'ian', 'ian', 'ian'],

  ['uã', null, null, 'uan'],
  [null, 'uan', 'uan', 'uan'],

  ['iẽ', 'iẽ', 'iẽ', 'ienn'],

  ['ẽ', null, null, 'enn'],
  [null, 'ə̃', null, 'enn'],
  [null, null, 'ɤ̃', 'enn'],

  ['yã', null, null, 'yan'],
  [null, 'yan', 'yan', 'yan'],

  ['yẽ', 'yẽ', 'yẽ', 'yenn'],
  ['on', null, null, 'on'],
  [null, 'õ', 'õ', 'on'],

  ['əȵ̍', null, null, 'en'],
  [null, 'ən', 'ən', 'en'],

  ['iȵ̍', null, null, 'in'],
  [null, 'in', 'in', 'in'],

  ['yȵ̍', null, null, 'yn'],
  [null, 'yn', 'yn', 'yn'],

  ['uəȵ̍', null, null, 'uen'],
  [null, 'uən', 'uən', 'uen'],

  ['ȵ̍', null, null, 'n'],
  ['aŋ', null, null, 'ang'],
  ['iaŋ', null, null, 'iang'],
  ['uaŋ', null, null, 'uang'],
  ['yaŋ', null, null, 'yang'],
  ['oŋ', 'oŋ', null, 'ong'],
  ['ioŋ', 'ioŋ', null, 'iong'],

  ['m̩', 'm̩', 'm̩', 'm'],
  ['n̩', 'n̩', 'n̩', 'n'],
]

export type AnyInitial = m1.Initial | m2.Initial | m3.Initial
export type AnyFinal = m1.Final | m2.Final | m3.Final
const senMap = new Map<AnyInitial, string>()
InitialConfig.forEach(([x, y, z, a]) =>
  [x, y, z].forEach((_) => _ !== null && senMap.set(_, a))
)
const ynMap = new Map<AnyFinal, string>()
FinalsConfig.forEach(([x, y, z, a]) =>
  [x, y, z].forEach((_) => _ !== null && ynMap.set(_, a))
)
export const getXPATuple = (ipaSen: AnyInitial, ipaYn: AnyFinal) => {
  // 只有湘音检字中有这个例外，可能对应单独 ɿ 音（对应 r 日）
  if (ipaSen === '' && ipaYn === m1.Finals.ɿ) {
    return ['ri', 'r', 'i'] as const
  }
  const sen = senMap.get(ipaSen)!
  const yn = ynMap.get(ipaYn)!
  return [sen + yn, sen, yn] as const
}
