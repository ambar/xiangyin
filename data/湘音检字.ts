import {changeTone} from './tones'
import {NormResult, QueryOptions} from './types'
import {AnyFinal, AnyInitial, toSianpinA} from './湘拼'
import json from './湘音检字.json'
import {Finals, Initials} from './湘音检字.meta'
export * from './湘音检字.meta'

const head = ['湘拼', '音標', '調號', '字甲', '字乙', '釋義'] as const

type DataItem = Record<typeof head[number], string> & {
  声: string
  韵: string
}

// 异体、多音字分组 16519 -> 13543
export const charGroup = new Map<string, DataItem[]>()

// 声母较长的排前面
const sortedInitials = Object.entries(Initials).sort(
  (a, b) => b[1].length - a[1].length
)

/** 分解原文 IPA */
export const ipa2senyn = (ipa: string): [string, string] => {
  let i = ''
  let f = ''
  // 先从零声母/韵母查起，因为 m̩（韵母）以 m（己有声母）开头
  const matchF = ipa in Finals
  if (matchF) {
    i = ''
    f = ipa
  } else {
    const matchSort = sortedInitials.find((x) => ipa.startsWith(x[1]))
    if (matchSort) {
      i = matchSort[0]
      f = i === '' ? ipa : ipa.replace(i, '')
    }
  }
  return [i, f]
}

export const items = json.map((x) => {
  const item = Object.fromEntries(x.map((v, i) => [head[i], v])) as DataItem
  const setIfNotSet = (char: string) => {
    if (char) {
      if (charGroup.has(char)) charGroup.get(char)!.push(item)
      else charGroup.set(char, [item])
    }
  }
  const [声, 韵] = ipa2senyn(item.音標)
  item.声 = 声
  item.韵 = 韵
  setIfNotSet(item.字甲)
  setIfNotSet(item.字乙)
  return item
})

export const query = (
  char: string,
  {pinyinType = 'IPA', toneType = 'CSToneNo'}: QueryOptions = {}
): NormResult[] => {
  const items = charGroup.get(char) || []
  return items.map((item) => {
    let tone: string | number = changeTone(item.調號, 'OctetToneNo', toneType)
    let {音標: 音, 声, 韵} = item
    if (pinyinType === 'XPA') {
      ;[音, 声, 韵] = toSianpinA(声 as AnyInitial, 韵 as AnyFinal)
    }
    return {音, 声, 韵, 调: tone, 释: item.釋義}
  })
}
