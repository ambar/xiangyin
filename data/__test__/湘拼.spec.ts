import {AnyFinal, AnyInitial, toSianpinA} from '../湘拼'
import * as xy from '../湘音检字'

const cases: [AnyInitial, AnyFinal][] = [
  [xy.Initials.p, xy.Finals.ã],
  [xy.Initials.k, xy.Finals.ai],
  ['', xy.Finals.ɿ],
  ['', xy.Finals.m̩],
]
test.each(cases)(`ipa2senyn: '%s,%s'`, (a, b) => {
  const xp = expect([a, b, toSianpinA(a, b)]).toMatchSnapshot()
})
