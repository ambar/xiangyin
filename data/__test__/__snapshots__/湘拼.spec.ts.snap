// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ipa2senyn: ',m̩' 1`] = `
Array [
  "",
  "m̩",
  Array [
    "m",
    "",
    "m",
  ],
]
`;

exports[`ipa2senyn: ',ɿ' 1`] = `
Array [
  "",
  "ɿ",
  Array [
    "ri",
    "r",
    "i",
  ],
]
`;

exports[`ipa2senyn: 'k,ai' 1`] = `
Array [
  "k",
  "ai",
  Array [
    "gai",
    "g",
    "ai",
  ],
]
`;

exports[`ipa2senyn: 'p,ã' 1`] = `
Array [
  "p",
  "ã",
  Array [
    "ban",
    "b",
    "an",
  ],
]
`;
